### PR TITLE
Protocol stack rework

### PR DESCRIPTION
**Pull request by scholzd**
Major rework to support variable sized headers (eg for IP/TCP options)
- removed header.lua, done per protocol now
- header template with default functions
- tcp options utility
- ethernet vlan support
- some bug fixes

As example, the stack is now able to correctly decode a eth (802.1q) - ip6 - udp - vxlan - eth - ip4 (with options) - udp - sflow buffer.

Probably breaks a lot of implementations in other forks.

### CODE REVIEW

File: lua/packet.lua [hunk 1]
```diff
@@ -461,33 +483,69 @@ function packetGet(self)
 	local namedArgs = {} 
 	local args = self:getArgs()
 	for i, v in ipairs(self:getHeaders()) do 
-		local _, member = getHeaderMember(args[i])
+		local _, member, _ = getHeaderData(args[i])
```
Comment (emmericp) [inline lua/packet.lua]: same as above: unused return value _

File: lua/packet.lua [hunk 2]
```diff
@@ -461,33 +483,69 @@ function packetGet(self)
...
 		namedArgs = mergeTables(namedArgs, v:get(member)) 
 	end 
 	return namedArgs 
 end
 
 --- Try to find out what the next header in the payload of this packet is.
 --- This function is only used for buf:get/buf:dump
+--- TODO support variable sized headers
```
Comment (emmericp) [inline lua/packet.lua]: is this TODO still accurate?

File: lua/packet.lua [hunk 3]
```diff
@@ -708,4 +817,89 @@ end
...
+
+---------------------------------------------------------------------------
+---- Protocol Stacks
+---------------------------------------------------------------------------
+
+pkt.getRawPacket = packetCreate()
+
+pkt.getEthernetPacket = packetCreate("eth")
+pkt.getEthPacket = pkt.getEthernetPacket
+pkt.getEthernetVlanPacket = packetCreate({"eth", "eth", subType = "vlan"})
```
Comment (emmericp) [inline lua/packet.lua]: would it be possible to allow {"eth", subType = "vlan"} if using the default name "eth"?

File: lua/packet.lua [hunk 4]
```diff
@@ -708,4 +817,89 @@ end
...
+	else 
+		return pkt.getIP6Packet(self) 
+	end 
+end   
+
+pkt.getArpPacket = packetCreate("eth", "arp")
+
+pkt.getIcmp4Packet = packetCreate("eth", "ip4", "icmp")
+pkt.getIcmp6Packet = packetCreate("eth", "ip6", "icmp")
+pkt.getIcmpPacket = function(self, ip4) ip4 = ip4 == nil or ip4 if ip4 then return pkt.getIcmp4Packet(self) else return pkt.getIcmp6Packet(self) end end
```
Comment (emmericp) [inline lua/packet.lua]: format into multiple lines similar to getIPPacket

File: lua/packet.lua [hunk 5]
```diff
@@ -708,4 +817,89 @@ end
...
+	end 
+end
+
+pkt.getSFlowPacket = packetCreate("eth", "ip4", "udp", {"sflow", subType = "ip4"}, "noPayload")
+
+pkt.getIpfixPacket = packetCreate("eth", "ip4", "udp", "ipfix")
+
+pkt.getLacpPacket = packetCreate('eth', 'lacp')
+
+pkt.getLongTestPacket = packetCreate({ "eth", subType="vlan"}, "ip6", "udp", "vxlan", { "eth", "innerEth" }, { "ip4", length = 20 }, {"udp", "innerUdp"},  {"sflow", subType = "ip4"}, "noPayload")
```
Comment (emmericp) [inline lua/packet.lua]: remove

File: lua/proto/ah.lua [hunk 6]
```diff
@@ -11,9 +10,8 @@
 ------------------------------------------------------------------------
 
 local ffi = require "ffi"
-local pkt = require "packet"
-
-require "headers"
+require"proto/template"
```
Comment (emmericp) [inline lua/proto/ah.lua]: space missing, also should be proto.template to be consistent

File: lua/proto/ip6.lua [hunk 7]
```diff
@@ -11,10 +10,10 @@
 ------------------------------------------------------------------------
 
 local ffi = require "ffi"
-local pkt = require "packet"
 
 require "utils"
-require "headers"
+require"proto/template"
```
Comment (emmericp) [inline lua/proto/ip6.lua]: missing space. also, please use proto.template instead to be consistent

File: lua/proto/ipfix.lua [hunk 8]
```diff
@@ -11,10 +10,10 @@
 ------------------------------------------------------------------------
 
 local ffi = require "ffi"
-local pkt = require "packet"
 
 require "utils"
-require "headers"
+require"proto/template"
```
Comment (emmericp) [inline lua/proto/ipfix.lua]: same, and probably in other files

File: lua/proto/tcp.lua [hunk 9]
```diff
@@ -470,11 +489,192 @@ function tcpHeader:getUrgentPointerString()
...
-	int = int or
-	self. = int
-end--]]
+
+------------------------------------------------------------------------------------
+---- TCP options
+------------------------------------------------------------------------------------
+
+do 
+	function createOptionsDict()
```
Comment (emmericp) [inline lua/proto/tcp.lua]: this should be local

File: lua/proto/tcp.lua [hunk 10]
```diff
@@ -470,11 +489,192 @@ function tcpHeader:getUrgentPointerString()
...
+			nop = 1,
+			mss = 2,
+			ws = 3,
+			ts = 8,
+		}
+		for k, v in pairs(mod.option) do
+			mod.option[v] = k
+		end
+	end
+	createOptionsDict()
```
Comment (emmericp) [inline lua/proto/tcp.lua]: seems pointless to declare a function and then call it immediately. just get rid of the function definition and call?

File: examples/pktgen.lua [hunk 11]
```diff
@@ -5,6 +5,7 @@ local stats  = require "stats"
 local log    = require "log"
 local memory = require "memory"
 local arp    = require "proto.arp"
+local pkt    = require "packet"
```
Comment (scholzd) [inline examples/pktgen.lua]: agreed

File: lua/packet.lua [hunk 12]
```diff
@@ -323,7 +335,7 @@ function packetCreate(...)
 	packet.calculateChecksums = packetCalculateChecksums(args)
 	
 	for _, v in ipairs(args) do
-		local header, member = getHeaderMember(v)
+		local header, member, _ = getHeaderData(v)
```
Comment (scholzd) [inline lua/packet.lua]: done

File: lua/packet.lua (hunk 1 above)
Comment (scholzd) [inline lua/packet.lua]: done

File: lua/packet.lua (hunk 2 above)
Comment (scholzd) [inline lua/packet.lua]: no, removed

File: lua/packet.lua (hunk 3 above)
Comment (scholzd) [inline lua/packet.lua]: yes

File: lua/packet.lua (hunk 4 above)
Comment (scholzd) [inline lua/packet.lua]: done for all occurences

File: lua/packet.lua (hunk 5 above)
Comment (scholzd) [inline lua/packet.lua]: done

File: lua/proto/tcp.lua (hunk 10 above)
Comment (scholzd) [inline lua/proto/tcp.lua]: done

File: lua/proto/ah.lua (hunk 6 above)
Comment (scholzd) [inline lua/proto/ah.lua]: all done

**Pull request merged**